File: linkList/4.palindromeLinkedList/go/isPalindrome.go
package _go

type ListNode struct {
	Val  int
	Next *ListNode
}

func isPalindrome(head *ListNode) bool {
	if head == nil || head.Next == nil {
		return true
	}
	// 快慢指针走到中点
	// 记录长度
	slow, quick := head, head
	for quick.Next != nil && quick.Next.Next != nil {
		quick = quick.Next.Next
		slow = slow.Next
	}
	var reHead *ListNode
	reHead = reverse(slow.Next)
	rNode, node := reHead, head
	for rNode != nil && node != nil && rNode.Val == node.Val {
		rNode = rNode.Next
		node = node.Next
	}
	return node == nil || rNode == nil
}

// 反转链表
func reverse(start *ListNode) *ListNode {
	var cur, pre, next *ListNode
	cur = start
	for cur != nil {
		next = cur.Next
		// 将指针指向上一个节点
		cur.Next = pre
		pre = cur
		// 到下一个节点
		cur = next
	}
	return pre
}

File: linkList/4.palindromeLinkedList/py/isPalindrome.py
from typing import Optional
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next
class Solution:
    def isPalindrome(self, head: Optional[ListNode]) -> bool:
        if head == None or head.next == None :
            return True
        slow = head
        quick = head
        while  not quick.next == None and not quick.next.next == None :
            quick = quick.next.next
            slow = slow.next
        rHead = self.reverse(slow.next)
        node = head
        rNode = rHead
        while rNode != None and node != None and node.val == rNode.val:
            rNode = rNode.next
            node = node.next
        return rNode == None or node == None
    def reverse(self,start:Optional[ListNode])->Optional[ListNode]:
        cur = start
        pre = None
        next = None
        while cur != None:
            next = cur.next
            cur.next = pre
            pre = cur
            cur = next
        return pre
            
        